iocore/net/SSLUtils.cc

```diff
@@ -100,6 +102,57 @@ static int ssl_vc_index = -1;
 static ink_mutex *mutex_buf      = nullptr;
 static bool open_ssl_initialized = false;
 
+void
+TLSKeyLogger::enable_keylogging_internal(const char *keylog_file)
+{
+  Debug("ssl_keylog", "Enabling TLS key logging to: %s.", keylog_file);
+  std::unique_lock lock{_mutex};
+  if (keylog_file == nullptr) {
+    _is_enabled = false;
+    Debug("ssl_keylog", "Received a nullptr for keylog_file: disabling keylogging.");
+    return;
+  }
+
+  _filestream = std::ofstream(keylog_file, std::ios_base::app);
+  if (!_filestream.is_open()) {
+    _is_enabled = false;
+    Error("Could not open keylog file: %s", keylog_file);
+    return;
+  }
+  _is_enabled = true;
+  Note("Opened %s for TLS key logging.", keylog_file);
+}
+
+void
+TLSKeyLogger::disable_keylogging_internal()
+{
+  std::unique_lock lock{_mutex};
+  if (_is_enabled) {
+    Note("Disabling TLS key logging.");
+  }
+  _is_enabled = false;
+  _filestream = std::ofstream{};
+
+  Debug("ssl_keylog", "TLS keylogging is disabled.");
+}
+
+void
+TLSKeyLogger::log(const char *line)
+{
+  std::shared_lock lock{_mutex};
+  if (!_is_enabled) {
+    return;
+  }
+
+  std::stringstream terminated_line;
+  terminated_line << line << std::endl;
+
+  // Writing the single, fully formed string to the file stream avoids
+  // concurrency issues with multiple writers to the keylog file.
+  instance()._filestream << terminated_line.str();
```

Comment:
I think there's a general problem here in that "Concurrent access to the same stream object may cause data races".